### ISSUE TITLE
How to train with long sequences using chunkwise feature of RetNet?

### ISSUE THREAD

**Issue by calliope-pro** [fkodom/yet-another-retnet#23]
Hello,

I am interested in training a model using the chunkwise feature of RetNet to handle long sequences.However, I couldn't find detailed instructions on how to do this in the documentation.
Could you please guide me on the best practices or steps to train a model with long sequences using the chunkwise in RetNet?

Thank you.

**Comment by fkodom** [comment]
Hi @calliope-pro!  For chunkwise usage, please check out [this section of the README](https://github.com/fkodom/yet-another-retnet#retnet).

The sequence returned from chunkwise formulation is the same as with parallel.  You can compute loss from those outputs:
```python
# Chunkwise usage
outputs = []  # container for collecting chunk-wise outputs
prev_states = []  # cache layer states after each step
chunk_size = 4  # number of tokens in each chunk
for idx in range(0, 16, chunk_size):
    out, prev_states = retnet.forward_chunkwise(
        x[:, idx : idx + chunk_size], idx, prev_states
    )
    outputs.append(out)
y_chunkwise = torch.cat(outputs, dim=1)
```

**Comment by calliope-pro** [comment]
I appreciate your reply.

I have another question regarding the memory usage. As the outputs array accumulates the results from each chunk, does this mean that the memory usage will increase linearly with the size of the outputs? Is there any way to reduce memory?

Thank you.

**Comment by fkodom** [comment]
@calliope-pro That's a good question.  Gradients will continue to accumulate, unless you `.detach()` the state Tensors between chunks.  But the state itself does not grow with sequence length.

A simple training step using chunkwise formulation could be:
```python
prev_states = []
chunk_size = 4
criterion = nn.CrossEntropyLoss()

# shape: (batch_size, seq_len)
x = torch.randint(0, 1000, (1, 16), device="cuda")
y = torch.randint(0, 1000, (1, 16), device="cuda")

loss = 0
for idx in range(0, 16, chunk_size):
    pred, prev_states = retnet.forward_chunkwise(
        x[:, idx : idx + chunk_size], idx, prev_states
    )
    labels = y[:, idx : idx + chunk_size]
    loss += criterion(rearrange(pred, "b n c -> (b n) c"), labels.flatten())
    prev_states = [state.detach() for state in prev_states]

loss.backward()
optimizer.step()
optimizer.zero_grad()
```

In this case, I believe `loss.backward()` would backprop through the model multiple times (once for each chunk).  

If you only wanted to backprop a single time, you could consider only computing loss based on the last chunk (and randomly truncate training sequences, so you don't overfit to larger sequence indices).  But that may be overcomplicating things. 🙃

**Comment by calliope-pro** [comment]
I appreciate your reply.

I tried `prev_states = [state.detach() for state in prev_states]`. However, I observed that this modification did not change the GPU memory usage before and after its application.
I am not using the `outputs`, but I am accumulating loss for each chunk. I've noticed that as more chunks are processed, the GPU memory usage keeps increasing. Is this increase in memory usage unavoidable when processing extremely long sequences?

Thank you.

**Comment by fkodom** [comment]
Short answer: I don't know for sure.  I haven't encountered this issue myself.

Longer answer: Maybe it's an issue with gradient accumulation?  In this loop, as I originally wrote it:
```python
loss = 0
for idx in range(0, 16, chunk_size):
    ...
    loss += criterion(rearrange(pred, "b n c -> (b n) c"), labels.flatten())
loss.backward()

optimizer.step()
optimizer.zero_grad()
```
The computation graph continues to grow with each chunk.  Instead, we could accumulate the gradients with each chunk, but still perform the optimizer step once after all loops are processed.

```python
for idx in range(0, 16, chunk_size):
    ...
    loss = criterion(rearrange(pred, "b n c -> (b n) c"), labels.flatten())
    loss.backward()

optimizer.step()
optimizer.zero_grad()
```

Disclaimer -- I haven't tried this explicitly.  Will hopefully have some time today to try this, and get a memory profile for you.  If so, I'll post an update here.

**Comment by calliope-pro** [comment]
I appreciate your reply.

Today, I tried the code you sent me.
```python
for idx in range(0, 16, chunk_size):
    ...
    loss = criterion(rearrange(pred, "b n c -> (b n) c"), labels.flatten())
    loss.backward()

optimizer.step()
optimizer.zero_grad()
```
This program seems to learn without increasing GPU memory.

Thank you so much.

**Comment by fkodom** [comment]
Glad to hear that!  Please feel free to re-open the issue if you have any other issues.

**Comment by calliope-pro** [comment]
As an postscript, I noticed both `[state.detach() for state in prev_states]` and `.backward(retain=True)` were required.
